### ISSUE TITLE
[intake] Implement the full backward pass for the MobileNetV1 CIFAR-10 tr

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#5524]
## Intake

**Idea:** Implement the full backward pass for the MobileNetV1 CIFAR-10 training script per ProjectOdyssey issue #3187, then validate by training one epoch with decreasing loss.

**Context:** Existing GitHub issue #3187 covers this work; scope strictly to it. The mesh should decompose into at most 3 dependency-ordered task-agent tasks, and the final PR must contain 'Closes #3187'.

Interview transcript and researched brief follow as comments.

**Comment by mvillmow** [comment]
**Q:** For the 'train one epoch with decreasing loss' acceptance criterion, does that mean running a full epoch over the real CIFAR-10 dataset (via the container, given the GLIBC 2.32+ constraint), or is the specced tiny-batch smoke test in tests/examples/test_mobilenetv1_train_step.mojo (loss2 < loss1 on a 2-sample batch) sufficient to satisfy validation?
**A:** Operator brief: scope strictly to ProjectOdyssey issue #3187 (full backward pass for the MobileNetV1 CIFAR-10 training script, examples/mobilenetv1-cifar10/train.mojo). If #3187 already has a plan with state:plan-go, follow it. Deliverables per the issue: forward pass with activation caching through all 13 depthwise separable blocks, full backward propagation through all 13 depthwise separable blocks, momentum velocity tensors, SGD+momentum parameter updates. Validation: run one training epoch on CIFAR-10 (a bounded subset is acceptable if a full epoch exceeds 60 minutes on CPU) and capture the log showing decreasing loss. Keep the workflow to at most 3 tasks assigned to pipeline.task-agent, dependency-ordered. The final PR must include 'Closes #3187'. _(via console)_

**Comment by mvillmow** [comment]
**Q:** Must the new Mojo smoke test actually execute and pass in CI/container before the PR merges, or is compile-plus-review acceptance enough (since Mojo runtime needs GLIBC 2.32+ and may not run on the local host)?
**A:** Operator brief: scope strictly to ProjectOdyssey issue #3187 (full backward pass for the MobileNetV1 CIFAR-10 training script, examples/mobilenetv1-cifar10/train.mojo). If #3187 already has a plan with state:plan-go, follow it. Deliverables per the issue: forward pass with activation caching through all 13 depthwise separable blocks, full backward propagation through all 13 depthwise separable blocks, momentum velocity tensors, SGD+momentum parameter updates. Validation: run one training epoch on CIFAR-10 (a bounded subset is acceptable if a full epoch exceeds 60 minutes on CPU) and capture the log showing decreasing loss. Keep the workflow to at most 3 tasks assigned to pipeline.task-agent, dependency-ordered. The final PR must include 'Closes #3187'. _(via console)_

**Comment by mvillmow** [comment]
**Q:** The issue accounts for 164 parameter slots but states both 110 and 136 trainable parameters in different places — should the implementation target exactly 110 velocity/trainable tensors as the code sketch (initialize_velocities → 110) specifies, treating the 136 mention as an error to ignore?
**A:** Operator brief: scope strictly to ProjectOdyssey issue #3187 (full backward pass for the MobileNetV1 CIFAR-10 training script, examples/mobilenetv1-cifar10/train.mojo). If #3187 already has a plan with state:plan-go, follow it. Deliverables per the issue: forward pass with activation caching through all 13 depthwise separable blocks, full backward propagation through all 13 depthwise separable blocks, momentum velocity tensors, SGD+momentum parameter updates. Validation: run one training epoch on CIFAR-10 (a bounded subset is acceptable if a full epoch exceeds 60 minutes on CPU) and capture the log showing decreasing loss. Keep the workflow to at most 3 tasks assigned to pipeline.task-agent, dependency-ordered. The final PR must include 'Closes #3187'. _(via console)_

**Comment by mvillmow** [comment]
## Researched brief

I have everything I need. The VGG16 script is a complete, shipping template with the exact structure (`compute_gradients`, `initialize_velocities`, a train loop wiring `CIFAR10Dataset.get_train_data()` → `compute_gradients` per batch), every backward primitive exists in `core/`, and the issue is `state:plan-go` with a fully-resolved plan.

---

# Researched Brief: Full Backward Pass for MobileNetV1 CIFAR-10 (#3187)

## Feasibility — High

This is a well-bounded port, not novel research. Three facts de-risk it:

1. **A complete working template exists.** `examples/vgg16_cifar10/train.mojo` (956 lines) is the only fully-implemented manual-backprop training script in the repo. It ships `compute_gradients` (forward+backward+SGD in one function, line 41), `initialize_velocities` (line 738), and a train loop that loads `CIFAR10Dataset.get_train_data()` and calls `compute_gradients` per batch (lines 845–906). MobileNetV1 mirrors this structure exactly — the only real difference is the depthwise-separable block topology.

2. **Every backward primitive already exists** in `src/projectodyssey/core/` — no new library code required:
   - `depthwise_conv2d_backward` (`conv.mojo:1074`) — the tricky per-channel piece is done.
   - `batch_norm2d_backward` (`normalization.mojo:519`) — returns a `Tuple`, *not* `GradientTriple`.
   - `global_avgpool2d_backward` (`pooling.mojo:764`), `conv2d_backward`, `relu_backward`, `linear_backward`, `cross_entropy_backward`.
   - `sgd_momentum_update_inplace` (`sgd.mojo:165`).

3. **The plan is approved.** Issue #3187 carries `state:plan-go` (plan approved by reviewer; implementer may proceed) plus a detailed, self-verified implementation plan. Design decisions are already pinned: 164 parameter slots, **110 trainable velocity tensors** (the "136" mention in the issue is a stale artifact — the code sketch `initialize_velocities → 110` is authoritative, per the operator brief and the 4 + 13×8 + 2 arithmetic), inline forward in `train.mojo` (not extending `model.mojo`), BN running-stats discarded this pass, per-block explicit code rather than a loop.

**Primary risk** is not correctness but the GLIBC 2.32+ / container constraint: the local host may not run the Mojo runtime, so the epoch-with-decreasing-loss log must be captured in the container (`just shell` → `just test-mojo`) or in CI. This is the reason validation is split into its own dependency-ordered task rather than folded into implementation.

## Prior art in this repo

| Artifact | Role |
| --- | --- |
| `examples/vgg16_cifar10/train.mojo` | Canonical manual-backprop template to mirror (compute_gradients + initialize_velocities + train loop). |
| `examples/mobilenetv1_cifar10/model.mojo` (1043 lines) | Forward-only model; 164 param slots; `DepthwiseSeparableBlock` struct at line 135; trainable fields `initial_conv_weights`, `initial_bn_gamma/beta`, per-block `dw_*`/`pw_*`, `fc_weights`, `fc_bias`. Untouched by this work. |
| `examples/mobilenetv1_cifar10/train.mojo:55–130` | The placeholder `train_epoch` printing a "to-be-implemented" banner — the target to replace. |
| `src/projectodyssey/core/{conv,normalization,pooling,loss}.mojo` | All backward primitives, already tested. |
| `examples/resnet18_cifar10/train.mojo` | Also a placeholder — confirms VGG16 is the *only* reference. |

## Proposed scope (strictly #3187)

Replace the placeholder body of `train_epoch` and add supporting functions in `examples/mobilenetv1_cifar10/train.mojo` only, plus one new smoke test `tests/examples/test_mobilenetv1_train_step.mojo`. No changes to `model.mojo`, `inference.mojo`, or any `core/` library. Deliverables per the issue: (a) forward pass with activation caching through the initial conv + 13 depthwise-separable blocks + global-avgpool + linear; (b) full manual backward producing gradients for the 110 trainable tensors; (c) 110 momentum velocity buffers + SGD-with-momentum in-place updates. **Validation:** run one epoch on CIFAR-10 (a bounded subset is acceptable if a full epoch exceeds 60 min on CPU) in the container and capture the log showing decreasing loss. Final PR includes `Closes #3187`.

Decomposed into **3 dependency-ordered tasks** (implementation → validation → PR), all assigned to `task-agent`.

## Ideated extensions (file as follow-ups, out of scope here)

1. **BN running-statistic updates + inference-mode fix.** This pass deliberately discards BN running mean/var (the backward path ignores them). Validation-mode `forward(training=False)` therefore still uses init-time `running_mean=0, running_var=1`, hurting eval accuracy. A follow-up should accumulate running stats during the forward pass so inference reflects the trained distribution.
2. **Refactor `compute_gradients` into forward / backward / update and share a loop-based block driver across VGG16, ResNet18, and MobileNetV1.** The issue acknowledges the single-mega-function SRP violation and per-block copy-paste as deliberate template-matching. A follow-up could extract a reusable manual-backprop harness, unblocking the ResNet18 placeholder as a bonus.

```yaml
apiVersion: telemachy/v1
metadata:
  name: mobilenetv1-backward-pass
  description: Implement, validate, and land the full MobileNetV1 CIFAR-10 backward pass per issue #3187
agents:
  - name: task-agent
    program: claude-code
teams:
  - name: implementation
    agents: [task-agent]
    tasks:
      - subject: Implement MobileNetV1 forward-cache, backward pass, and SGD-momentum updates
        description: >
          In examples/mobilenetv1_cifar10/train.mojo ONLY (plus one new test file), replace the
          placeholder train_epoch body following the state:plan-go plan on issue #3187 and mirroring
          the working template in examples/vgg16_cifar10/train.mojo (compute_gradients at line 41,
          initialize_velocities at line 738). Add: (1) an inline forward pass caching every BN-input,
          BN-output, ReLU pre-activation, and per-block input across the initial conv + 13
          depthwise-separable blocks + global avgpool + linear; (2) a full manual backward pass using
          existing core primitives — depthwise_conv2d_backward (conv.mojo:1074), batch_norm2d_backward
          (normalization.mojo:519, returns a Tuple not GradientTriple), global_avgpool2d_backward
          (pooling.mojo:764), conv2d_backward, relu_backward, linear_backward, cross_entropy_backward —
          producing gradients for the 110 trainable tensors; (3) initialize_velocities returning EXACTLY
          110 zero velocity buffers in the documented order (4 initial-conv + 13*8 block + 2 fc; the
          "136 trainable" mention in the issue is stale — 110 is authoritative); (4) in-place
          sgd_momentum_update_inplace (sgd.mojo:165) updates. Do NOT modify model.mojo, inference.mojo,
          or any core/ library code. BN running stats are discarded this pass (documented as follow-up).
          Add tests/examples/test_mobilenetv1_train_step.mojo asserting: velocity count == 110, two
          successive losses on a fixed 2-sample batch are finite-positive and strictly decreasing, at
          least one parameter changed, and post-train forward returns (2,10) logits. Verify: the file
          compiles (mojo build) and, where the container/GLIBC 2.32+ runtime is available, the smoke
          test executes and prints PASSED.
        assign_to: task-agent
        blocked_by: []
      - subject: Validate one CIFAR-10 training epoch shows decreasing loss
        description: >
          Using the implementation from the prior task, run one training epoch of the MobileNetV1
          CIFAR-10 script on real CIFAR-10 data inside the container (just shell then just test-mojo /
          the example's train entrypoint, since the Mojo runtime needs GLIBC 2.32+ and may not run on
          the host). A bounded subset of the training set is acceptable if a full epoch exceeds 60
          minutes on CPU — document the subset size if used. Capture the epoch log demonstrating batch
          loss decreasing over the epoch. If loss fails to decrease or produces NaN/inf, diagnose via
          the smoke test and hand corrections back before proceeding. Save the captured log for the PR.
        assign_to: task-agent
        blocked_by:
          - Implement MobileNetV1 forward-cache, backward pass, and SGD-momentum updates
      - subject: Open PR closing issue #3187 with the decreasing-loss evidence
        description: >
          Create a feature branch (3187-mobilenetv1-backward-pass), run just precommit (never
          --no-verify), and open a pull request whose body contains 'Closes #3187'. Include in the PR
          body: a summary of the forward-cache/backward/SGD implementation, the 110-trainable-tensor
          decision, confirmation that only train.mojo and the new test file changed, and the captured
          one-epoch decreasing-loss log from the validation task as verification evidence. Enable
          auto-merge with rebase (gh pr merge --auto --rebase).
        assign_to: task-agent
        blocked_by:
          - Validate one CIFAR-10 training epoch shows decreasing loss
```

**Comment by mvillmow** [comment]
Registered epic #5528 (key `homericintelligence-projectodyssey-5528`).

**Comment by mvillmow** [comment]
Target ProjectOdyssey #3187 (Full backward pass for MobileNetV1) is CLOSED; the tracked work (epic #5528: sub-tasks #5525/#5526/#5527) is all CLOSED. Done.
